components/compositing/compositor.rs

```diff
@@ -862,20 +867,69 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
         }
     }
 
+    /// Create a WebRender display list for the root pipeline in the given transaction.
+    /// This display list contains a full-screen iframe with the root content pipeline
+    /// with any pinch zoom transform applied to it.
+    fn set_root_display_list(&self, transaction: &mut Transaction) {
+        let root_content_pipeline = match self.root_content_pipeline.id {
+            Some(id) => id,
+            None => return,
+        };
+
+        let root_pipeline = WebRenderPipelineId(0, 0);
+        transaction.set_root_pipeline(root_pipeline);
+
+        let mut builder = webrender_api::DisplayListBuilder::new(root_pipeline);
+        let viewport_size = LayoutSize::new(
+            self.embedder_coordinates.get_viewport().width() as f32,
+            self.embedder_coordinates.get_viewport().height() as f32,
+        );
+        let zoom_factor = self.pinch_zoom_level();
+        let viewport_rect = LayoutRect::new(LayoutPoint::zero(), viewport_size);
+        let zoom_reference_frame = builder.push_reference_frame(
+            LayoutPoint::zero(),
+            SpatialId::root_reference_frame(root_pipeline),
+            TransformStyle::Flat,
+            PropertyBinding::Value(Transform3D::scale(zoom_factor, zoom_factor, 1.)),
+            ReferenceFrameKind::Transform {
+                is_2d_scale_translation: true,
+                should_snap: true,
+            },
+        );
+
+        builder.push_iframe(
+            viewport_rect,
+            viewport_rect,
+            &SpaceAndClipInfo {
+                spatial_id: zoom_reference_frame,
+                clip_id: ClipId::root(root_pipeline),
+            },
+            root_content_pipeline.to_webrender(),
+            true,
+        );
+        let built_display_list = builder.finalize();
+        transaction.set_display_list(
+            WebRenderEpoch(0),
```

Comment:
Why do we use zero value for epoch here?

Reply:
The epoch value doesn't really matter here as WebRender continues to generate frames and update the scene no matter what we pass. I've added a comment explaining this.